### PR TITLE
Added a barrier after every gradient accumulation step.

### PR DESCRIPTION
**Pull request by taylanbil**
That is, every time we do a forward pass without a backward pass.

Also removed the null assignment to `tracker` in the main function.

### CODE REVIEW

**Comment by taylanbil** [comment]
This was requested in https://github.com/pytorch/xla/issues/1367

**Review by jysohn23 — APPROVED** [review]

**Comment by taylanbil** [comment]
Let me clarify the flow here a little bit:

1. We mark step every time we yield `samples` from `loader` in [this line](https://github.com/pytorch-tpu/fairseq/blob/tpu/train.py#L398)
2. With `--update-freq=1` (default setting), we do barrier + forward + backward (I mean optimizer step here) + barrier + ... loop
3. With `--update-freq=10`, we do `barrier + forward1 + forward2 + ... + forward10 + backward + barrier + ...
4. With this pr, we will do `barrier (from yield) + forward1 + barrier (from this pr) + forward2 + barrier (from this pr) + ... + barrier (from this pr) + forward10 + backward + barrier (from yield)

File: fairseq/trainer.py [hunk 1]
```diff
@@ -342,10 +344,17 @@ def maybe_no_sync():
                 else:
                     raise e
 
+            if self.xla and i < len(samples)-1:
+                # tpu-comment: every xla operation before marking step is
+                # appended to the IR graph, and processing too many batches
+                # before marking step can lead to OOM errors.
+                # To handle gradient accumulation use case, we explicitly
+                # mark step here for every forward pass without a backward pass
+                xm.mark_step()
```
Comment (dlibenzi) [inline fairseq/trainer.py:353]: is the sample iteration above inside of the ParallelLoader one?
Because the ParallelLoader already barrier.

Comment (taylanbil) [reply fairseq/trainer.py:353]: parallel loader yields `samples`. with `--update-freq=10`, it will yield 10 batches per step. I tried to list out how barriers would look like in point 4 inside https://github.com/pytorch-tpu/fairseq/pull/22#issuecomment-557223044

**Review by dlibenzi — APPROVED** [review]
